src/audio/drc/ref_sin_fixpt.c

```diff
@@ -0,0 +1,319 @@
+/*
+ * File: ref_sin_fixpt.c
+ *
+ */
+ // SPDX - License - Identifier: BSD - 3 - Clause
+ //
+ //Copyright(c) 2021 Intel Corporation.All rights reserved.
+ //
+ //Author : Shriram Shastry <malladi.sastry@linux.intel.com>
+
+/* Include Files */
+#include "ref_sin_fixpt.h"
+#include <string.h>
+#include <stdint.h>
+#include "typedef.h"
+#include <errno.h>
+
+
+/* Function Definitions */
+/*
+ * function [cdcSinTh] = ref_sine_fixpt(thRadFxp)
+ * Use Q12.20-bit quantized inputs and number of iterations to 10.
+ *  Compare the fixed-point cordicsin function results to the
+ *  results of the double-precision sin function.
+ * +--------------+---------+----------+----------+---------+
+ * |              |thRadFxp |cdcSinTh  |QthRadFxp |QcdcSinTh|
+ * +--------------+---------+----------+----------+---------+
+ * |    WordLength| 31      |  32      |12.20     | 3.29    |
+ * |FractionLength| 20      |  29      |          |         |
+ * +--------------+---------+----------+----------+---------+
+ *  THD+N = 29*6 = 174
+ * Arguments    : const int32_t thRadFxp[1024]
+ *                int32_t cdcSinTh[1024]
+ * Return Type  : void
+ */
+void ref_sine_fixpt(const int32_t thRadFxp[1024], int32_t cdcSinTh[1024])
+{
+  static const int32_t inpLUT[10] = { 421657428, 248918915, 131521918, 66762579, 33510843, 16771758, 8387925, 4194219, 2097141, 1048575 };
+
+  int32_t b_idx;
+  int32_t b_yn;
+  int32_t idx;
+  int32_t th;
+  int32_t xn;
+  int32_t xtmp;
+  int32_t ytmp;
+  int32_t z;
+  boolean_t negate;
+
+  for (idx = 0; idx < 1024; idx++) {
+    th = thRadFxp[idx];
+    if (th > 1647099) {                               // +20
+      if (((th - 3294199) & 1073741824) != 0) {
+        z = (th - 3294199) | -1073741824;
+      } else {
+        z = (th - 3294199) & 1073741823;
+      }
+
+      if (z <= 1647099) {
+        th = z;
+        negate = 1U;
+      } else {
+        if (((th - 6588397) & 1073741824) != 0) {
+          th = (th - 6588397) | -1073741824;
+        } else {
+          th = (th - 6588397) & 1073741823;
+        }
+
+        negate = 0U;
+      }
+    } else if (th < -1647099) {                         // -20
+      if (((th + 3294199) & 1073741824) != 0) {
+        z = (th + 3294199) | -1073741824;
+      } else {
+        z = (th + 3294199) & 1073741823;
+      }
+
+      if (z >= -1647099) {
+        th = z;
+        negate = 1U;
+      } else {
+        if (((th + 6588397) & 1073741824) != 0) {
+          th = (th + 6588397) | -1073741824;
+        } else {
+          th = (th + 6588397) & 1073741823;
+        }
+
+        negate = 0U;
+      }
+    } else {
+      negate = 0U;
+    }
+
+    th <<= 9;
+    if ((th & 1073741824) != 0) {
+      z = th | -1073741824;
+    } else {
+      z = th & 1073741823;
+    }
+
+    b_yn = 0;
+    xn = 326016644;
+    xtmp = 326016644;
+    ytmp = 0;
+    for (b_idx = 0; b_idx < 10; b_idx++) {
+      if (z < 0) {                              // negative LookUp table
```

Comment:
this is definitely not the easiest to read, we should look at some methods to break the up, either through switch cases, gotos or sub functions